examples/halo_test/halo_test.f90

```diff
@@ -349,93 +393,73 @@ subroutine test_div_haloZ()
-      call alloc_z(w3, global)
-
       call transpose_y_to_z(u2, u3)
       call transpose_y_to_z(v2, v3)
       call transpose_y_to_z(w2, w3)
 
-      call alloc_x(div4, global)
-      call alloc_y(wk2, global)
-      call alloc_z(wk3, global)
-
       ! du/dx
 #ifdef HALO_GLOBAL
       call update_halo(u3, uh, 1, opt_global=.true., opt_pencil=3)
+      call update_halo(v3, vh, 1, opt_global=.true., opt_pencil=3)
       ifirst = zstart(1); ilast = zend(1)
       jfirst = zstart(2); jlast = zend(2)
 #else
       call update_halo(u3, uh, 1, opt_pencil=3)
+      call update_halo(v3, vh, 1, opt_pencil=3)
       ifirst = 1; ilast = zsize(1)
       jfirst = 1; jlast = zsize(2)
 #endif
       kfirst = 2; klast = zsize(3) - 1
 
       call test_halo_size(uh, nx_expected, ny_expected, nz_expected, "Z:u")
-
-      do j = jfirst, jlast
-         do i = ifirst, ilast
-            do k = kfirst, klast
-               wk3(i, j, k) = uh(i + 1, j, k) - uh(i - 1, j, k)
-            end do
-         end do
-      end do
-
-      ! dv/dy
-#ifdef HALO_GLOBAL
-      call update_halo(v3, vh, 1, opt_global=.true., opt_pencil=3)
-#else
-      call update_halo(v3, vh, 1, opt_pencil=3)
-#endif
-
       call test_halo_size(vh, nx_expected, ny_expected, nz_expected, "Z:v")
 
+      !$acc kernels default(present)
       do j = jfirst, jlast
          do i = ifirst, ilast
             do k = kfirst, klast
-               wk3(i, j, k) = wk3(i, j, k) + vh(i, j + 1, k) - vh(i, j - 1, k)
-            end do
-         end do
-      end do
-
-      ! dw/dz
-      do j = jfirst, jlast
-         do i = ifirst, ilast
-            do k = kfirst, klast
-               wk3(i, j, k) = wk3(i, j, k) + w3(i, j, k + 1) - w3(i, j, k - 1)
+               wk3(i, j, k) = uh(i + 1, j, k) - uh(i - 1, j, k) &
+                            + vh(i, j + 1, k) - vh(i, j - 1, k) &
+                            + w3(i, j, k + 1) - w3(i, j, k - 1)
             end do
          end do
       end do
+      !$acc end kernels
 
       call transpose_z_to_y(wk3, wk2)
-      call transpose_y_to_x(wk2, div4)
+      call transpose_y_to_x(wk2, div3)
 
       ! Compute error
-      call check_err(div4, "Z")
+      call check_err(div3, "Z")
 
-      deallocate (uh, vh, wk2, wk3)
+      deallocate (uh, vh)
    end subroutine test_div_haloZ
-
+   !=====================================================================
+   ! Check the difference between halo and transpose divergence
+   !=====================================================================
    subroutine check_err(divh, pencil)
 
+      implicit none
+      
       real(mytype), dimension(:, :, :), intent(in) :: divh
       character(len=*), intent(in) :: pencil
-
       real(mytype), dimension(:, :, :), allocatable :: tmp
+      real(mytype) :: divmag, error
+#if defined(_GPU)
+      attributes(device) :: tmp
+#endif
 
-      real(mytype) :: divmag
-
-      ! XXX: The Intel compiler SEGFAULTs if the array difference is computed inplace
```

Comment:
I think useful to keep comment in case someone tries to remove tmp array as it's not (obviously) required...

Reply:
Resolved by 5ab1f79